hypershift-operator/controllers/scheduler/dedicated_request_serving_nodes.go

```diff
@@ -257,3 +272,366 @@ func (r *DedicatedServingComponentScheduler) Reconcile(ctx context.Context, req
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type placeholderLister struct {
+	getClusterSizingConfiguration func(context.Context) (*schedulingv1alpha1.ClusterSizingConfiguration, error)
+	getDeployment                 func(context.Context, types.NamespacedName) (*appsv1.Deployment, error)
+	listDeployments               func(context.Context, ...client.ListOption) (*appsv1.DeploymentList, error)
+	listNodes                     func(context.Context, ...client.ListOption) (*corev1.NodeList, error)
+}
+
+type placeholderCreator struct {
+	client kubernetes.Interface
+	*placeholderLister
+}
+
+func (r *placeholderCreator) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+	logger.Info("Reconciling")
+
+	config, err := r.getClusterSizingConfiguration(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	deployment, err := r.reconcile(ctx, config)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if deployment != nil {
+		_, err := r.client.AppsV1().Deployments(placeholderNamespace).Apply(ctx, deployment, metav1.ApplyOptions{FieldManager: "PlaceholderScheduler"})
+		return reconcile.Result{Requeue: true}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *placeholderCreator) reconcile(
+	ctx context.Context,
+	config *schedulingv1alpha1.ClusterSizingConfiguration,
+) (*appsv1applyconfigurations.DeploymentApplyConfiguration, error) {
+	var configValid bool
+	for _, condition := range config.Status.Conditions {
+		if condition.Type == schedulingv1alpha1.ClusterSizingConfigurationValidType && condition.Status == metav1.ConditionTrue {
+			configValid = true
+			break
+		}
+	}
+	if !configValid {
+		// we can't deliver placeholders unless we have a valid configuration; we'll re-trigger when
+		// the configuration object changes and can process deployments then
+		return nil, nil
+	}
+
+	for _, sizeClass := range config.Spec.Sizes {
+		if sizeClass.Management != nil && sizeClass.Management.Placeholders != 0 {
+			deployments, err := r.listDeployments(ctx, client.HasLabels{PlaceholderLabel}, client.MatchingLabels{hostedclustersizing.HostedClusterSizeLabel: sizeClass.Name})
+			if err != nil {
+				return nil, err
+			}
+
+			if len(deployments.Items) >= sizeClass.Management.Placeholders {
+				// we already have all the placeholders we need, nothing to do
+				return nil, nil
+			}
+
+			// we need more placeholders, add one and requeue
+			nodes, err := r.listNodes(ctx, client.HasLabels{HostedClusterNameLabel, OSDFleetManagerPairedNodesLabel})
+			if err != nil {
+				return nil, err
+			}
+
+			pairedNodesAssignedToHostedClusters := sets.Set[string]{}
+			for _, node := range nodes.Items {
+				pairedNodesAssignedToHostedClusters.Insert(node.ObjectMeta.Labels[OSDFleetManagerPairedNodesLabel])
+			}
+
+			return newDeployment(placeholderNamespace, sizeClass.Name, len(deployments.Items), pairedNodesAssignedToHostedClusters.UnsortedList()), nil
```

Comment:
len(deployments.Items) may not always be the right index to use. If somehow a placeholder pod was deleted, then you'll likely try to create a dup. I think it might make more sense to use a generated suffix (as in .metadata.generateName).